pkg/controller/common/helpers.go

```diff
@@ -658,3 +661,120 @@ func dedupePasswdUserSSHKeys(passwdUser ign2types.PasswdUser) ign2types.PasswdUs
 
 	return passwdUser
 }
+
+// CalculateConfigFileDiffs compares the files present in two ignition configurations and returns the list of files
+// that are different between them
+func CalculateConfigFileDiffs(oldIgnConfig, newIgnConfig *ign3types.Config) []string {
+
+	// Go through the files and see what is new or different
+	oldFileSet := make(map[string]ign3types.File)
+	for _, f := range oldIgnConfig.Storage.Files {
+		oldFileSet[f.Path] = f
+	}
+	newFileSet := make(map[string]ign3types.File)
+	for _, f := range newIgnConfig.Storage.Files {
+		newFileSet[f.Path] = f
+	}
+	diffFileSet := []string{}
+
+	// First check if any files were removed
+	for path := range oldFileSet {
+		_, ok := newFileSet[path]
+		if !ok {
+			// debug: remove
+			glog.Infof("File diff: %v was deleted", path)
+			diffFileSet = append(diffFileSet, path)
+		}
+	}
+
+	// Now check if any files were added/changed
+	for path, newFile := range newFileSet {
+		oldFile, ok := oldFileSet[path]
+		if !ok {
+			// debug: remove
+			glog.Infof("File diff: %v was added", path)
+			diffFileSet = append(diffFileSet, path)
+		} else if !reflect.DeepEqual(oldFile, newFile) {
+			// debug: remove
+			glog.Infof("File diff: detected change to %v", newFile.Path)
+			diffFileSet = append(diffFileSet, path)
+		}
+	}
+	return diffFileSet
+}
+
+// GetNewestCertificatesFromPEMBundle breaks a pem-encoded bundle out into its component certificates
+func GetCertificatesFromPEMBundle(pemBytes []byte) ([]*x509.Certificate, error) {
+	var certs []*x509.Certificate
+
+	// There can be multiple certificates in the file
+	for {
+
+		// Decode a block to parse
+		block, rest := pem.Decode(pemBytes)
+
+		// Once we get no more blocks, we've read all the certs
+		if block == nil {
+			break
+		}
+
+		// Right now we just care about certificates, not keys
+		if block.Type == "CERTIFICATE" {
+
+			cert, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				// This isn't fatal, *this* cert could just be junk, next one could be okay
+				glog.Warningf("Failed to parse certificate: %v", err.Error())
+			} else {
+				certs = append(certs, cert)
+			}
+		}
+
+		// Keep reading from where we left off
+		pemBytes = rest
+	}
+
+	return certs, nil
+}
+
+// GetIgnitionFileDataByPath retrieves the file data for a specified path from a given ignition config
+func GetIgnitonFileDataByPath(config *ign3types.Config, path string) ([]byte, error) {
+
```

Comment:
nit: extra newline

Reply:
I am...so sorry about my love of extra newlines. I will clean all these up.